[IMP] account: Simplify tax_audit computation

rousseldenis opened the pull request:
Tax Audit field computation was very odd to understand and reasons code have been added is not obvious.

This is a proposal of simplification in order to use the Tax Negate field in account tags in order to determine the sign value.

Note: I've found the tax negate field help : https://github.com/odoo/odoo/blob/14.0/addons/account/models/account_account_tag.py#L15

I think this is what code does now and surely wanted when refactor done by @smetl there https://github.com/odoo/odoo/commit/7076b4f4d0d933d93b24e8e4c7cf21ef0b0008e5

Tasks:

- [x] Simplify tax audit computation
- [x] Fix tests in point of sale - Create refund taxes with Tax Negate == True
- [x] Fix mixed invoices with normal lines and discount lines - negative quantity or price (those ones have wrong tax tag affected - Note: In a miscellaneous operation, behavior is correct)

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/88859).

rousseldenis commented:
@jbw-odoo I need your advice here as you did the last modifications at that level.

IMHO, the code become too tough to understand. I still need to add some tests but I think this is the good assertion in order to determine the sign of tax audit. You should not rely on journal but on account properties.

 @smetl

rousseldenis commented:
opw-2825159

jbw-odoo commented:
@oco-odoo this what you added here: https://github.com/odoo/odoo/commit/ea4ef3d20614a5f6ee929f36a804ecd1a55d12c7

rousseldenis commented:
@oco-odoo If you can provide me a test case, I can add it here in order to avoid regression. Thanks

rousseldenis commented:
This one solves the third task in PR description:

https://github.com/odoo/odoo/pull/88859/commits/3e02aa3da3a1350a9a99eed69b4982e7ce23ba48

IMHO the behavior split between 'entry' move lines and invoice ones should be avoided as possible as it could lead to wrong computations.

Everything should be able to be done in entries with the same behavior as in invoices.

rousseldenis commented:
@sanrav @sbidoul @regispirard

Levizar commented on addons/point_of_sale/tests/common.py:
```diff
@@ -352,17 +352,30 @@ def _create_taxes(cls):
+        def create_tag(name, negate=False):
             return cls.env['account.account.tag'].create({
                 'name': name,
                 'applicability': 'taxes',
-                'country_id': cls.env.company.country_id.id
+                'country_id': cls.env.company.country_id.id,
+                'tax_negate': negate,
+                'tax_report_line_ids': [
+                    (0, 0, {
+                        'name': name,
+                        'tag_name': name,
+                        'report_id': cls.tax_report.id,
+                        'sequence': 10,
+                    })
+                ]
             })
+        cls.tax_report = cls.env['account.tax.report'].create({
+            'name': "Tax report",
+            'country_id': cls.company_data['company'].country_id.id,
+        })
 
         cls.tax_tag_invoice_base = create_tag('Invoice Base tag')
         cls.tax_tag_invoice_tax = create_tag('Invoice Tax tag')
-        cls.tax_tag_refund_base = create_tag('Refund Base tag')
-        cls.tax_tag_refund_tax = create_tag('Refund Tax tag')
+        cls.tax_tag_refund_base = create_tag('Refund Base tag', True)
+        cls.tax_tag_refund_tax = create_tag('Refund Tax tag', True)
```
Isn't it overkill to test the `tax_audit_string` in the PoS ?  You can probably just make similar test just by creating a misc move.

Aside of that, if you really want to test it, you will need to get the tag created from the report line and not to create the report line from the tag. (Because the report.line creates 2 tags: a positive and a negative by itself and that's how we are really using those)

rousseldenis replied:
@Levizar I did it because some changes in account were done specifically for POS (see the diff where I remove the hook https://github.com/odoo/odoo/pull/88859/files#diff-1e3bd6be3bfb83a37ec9fb800ce8b1c95afe0be90ff792874ae7299c320a2f6eL3878).

As it's also overridden in POS code, I wanted to show that nothing has impacted (unless the wanted behavior)

Levizar commented on addons/account/tests/test_account_move_in_invoice.py:
```diff
@@ -619,11 +637,12 @@ def test_in_invoice_line_onchange_accounting_fields_1(self):
             },
             {
                 **self.tax_line_vals_2,
-                'price_unit': 250.0,
-                'price_subtotal': 250.0,
-                'price_total': 250.0,
-                'amount_currency': 250.0,
-                'debit': 250.0,
+                'price_unit': -75.0,
+                'price_subtotal': -75.0,
+                'price_total': -75.0,
+                'amount_currency': -75.0,
+                'credit': 75.0,
+                'debit': 0.0,
```
Seeing this, I realized that it's not just a small "refactoring" but a whole change of behavior which will have different results. This means that it may be refused.

rousseldenis replied:
Yes I realize it.

But the current behavior is not correct as a negative line for tax should be generated with the good tag (negative one).

You can reproduce on runbot:

Load a localization with positive and negative tags (l10n_lu for instance).
Create an OD with a sale line (tax line automatically generated, + tag on revenue account), a rebate line (tax line automatically generated, - tag on the other revenue account).

Create an Invoice, with both lines. Check the generated entries.
Only one line has been generated for tax (with the + tag).

rousseldenis replied:
IMHO, the OD has the correct behaviour.

Levizar reviewed (changes requested):
It's a good idea to unify behaviors but I don't know if that will be ok.

rousseldenis commented on addons/account/tests/common.py:
```diff
@@ -491,6 +492,15 @@ def get_xml_tree_from_string(self, xml_tree_str):
         '''
         return etree.fromstring(xml_tree_str)
 
+    @classmethod
+    def _get_tax_audit_string(cls, move, tag, amount):
+        audit_string = ''
+        separator = '        '
+        for report_line in tag.tax_report_line_ids:
+            audit_string += separator if audit_string else ''
+            audit_string += report_line.tag_name + ': ' + formatLang(cls.env, amount, currency_obj=move.currency_id)
+        return audit_string
+
```
Indeed, I thought about that and will change it

rousseldenis commented:
@Levizar Are you in charge of the OPW ?

Levizar commented:
> @Levizar Are you in charge of the OPW ?

No (at least not atm). 
I was just automatically assigned for a first round of review.

oco-odoo commented:
@rousseldenis 
Stop me if I'm wrong, but it seems to me you're not fixing anything with this, but rather totally refactoring the current implementation to have something you consider clearer. If this is the case, there is no way we merge this into a stable version, for obvious reasons. 

The fact we inverse the tags on misc operations under some circumstances is indeed something we had to do but did not like to. This is why, in 15.0, I refactored the way the tax report computes its amounts to make use of a new boolean field on account.move.line telling whether of not we need to invert the balance of the line before applying the tag. With this, we keep the same tag signs on both invoices or "equivalent" misc operations. 
See: https://github.com/odoo/odoo/commit/6576e1388e0a3d2a822a768c6c563769afb392a1


If you're trying to fix something more precise; let's discuss the bug, isolate it and solve it with minimal changes, as suits a stable version.

rousseldenis commented:
> @rousseldenis Stop me if I'm wrong, but it seems to me you're not fixing anything with this, but rather totally refactoring the current implementation to have something you consider clearer. If this is the case, there is no way we merge this into a stable version, for obvious reasons.
> 
> The fact we inverse the tags on misc operations under some circumstances is indeed something we had to do but did not like to. This is why, in 15.0, I refactored the way the tax report computes its amounts to make use of a new boolean field on account.move.line telling whether of not we need to invert the balance of the line before applying the tag. With this, we keep the same tag signs on both invoices or "equivalent" misc operations. See: [6576e13](https://github.com/odoo/odoo/commit/6576e1388e0a3d2a822a768c6c563769afb392a1)
> 
> If you're trying to fix something more precise; let's discuss the bug, isolate it and solve it with minimal changes, as suits a stable version.

If you have a workaround for:
- Correct tag computation for a negative amount in a misc operation

I will take it as indeed for now, it's quite a huge thing for that bug. Thanks

oco-odoo commented:
@rousseldenis 

So, if I get it well (if not, please give a concrete example), you're setting the tags by hand directly on a misc operation ? Indeed, in this case, they sometimes need to be inverted so that the balance in the tax report is correct. So the fact that the report does + if you put a + tag is not guaranteed, for design reasons (which we fixed in 15.0, as I said). What is guaranteed though (or what should be, if there is a bug) is that the amounts shown in the tax_audit field are the same as the ones taken into account in the report. 

For manual adjustments like this, we implemented the tax adjustment wizard, which directly takes in charge the inversion of tax tags when it's needed ; did you consider using it ?

rousseldenis commented:
> @rousseldenis
> 
> So, if I get it well (if not, please give a concrete example), you're setting the tags by hand directly on a misc operation ? Indeed, in this case, they sometimes need to be inverted so that the balance in the tax report is correct. So the fact that the report does + if you put a + tag is not guaranteed, for design reasons (which we fixed in 15.0, as I said). What is guaranteed though (or what should be, if there is a bug) is that the amounts shown in the tax_audit field are the same as the ones taken into account in the report.
> 
> For manual adjustments like this, we implemented the tax adjustment wizard, which directly takes in charge the inversion of tax tags when it's needed ; did you consider using it ?

In fact, it's not about a tax adjustement really.

It's a correction of POS sale orders done through a misc operation.

So, in fact, we are expecting the behavior should be the same as for normal invoices (as everything should be done through a misc operation - that's common for accounting people).

rousseldenis commented:
Here is a normal invoice just done on runbot:

- One line of sale
- On line of rebate
- Both with taxes

![image](https://user-images.githubusercontent.com/19529533/165923515-bea6a350-fcd8-4d28-be3b-9899d54f48ed.png)

You see the tax report:

![image](https://user-images.githubusercontent.com/19529533/165923775-b9b639fc-b400-4e46-a58f-4b14ca2dcf5b.png)

Amounts are correct with the good sign (we could argue that the negative amount should have a tag '-701').

Here is the misc operation that should do the same:

![image](https://user-images.githubusercontent.com/19529533/165924555-4f180b17-0ef1-4e51-b92d-570bde8e8660.png)

The corresponding tax report lines:

![image](https://user-images.githubusercontent.com/19529533/165924626-d9d70bb6-1528-4cbe-8747-9f2347c6f285.png)



I see something has changed since the begin of this thread.

Amounts signs are correct.

So, you can close this.

But IMHO, the misc operation tax sign is not correct for credit lines (should be '+701') and for invoice rebate line (should be '-701').

What do you think ?

oco-odoo commented:
Tag inversion on  misc operations was indeed confusing in 13/14, yet it was intended, and the report (and tax audit string) computation was correct. Since the report is alright, we won't change anything about that in those versions (for stability). 
In v15, the implementation has changed, and the use of a new field on move lines allows avoiding this explicit inversion.
